### PR TITLE
Allow accessing data by paths

### PR DESCRIPTION
**Pull request by ftk**
#17
Rationale:
- HTTP clients can infer the filename by path (without even starting the download), so there is no need for `?filename=`
- Some files can refer to other files in the same directory. For example, the file is a playlist and media player now can seamlessly load other files it refers to using http.

This PR adds these endpoints:
`GET /data/infohash/<infohash in hex>/<display path of file declared in torrent info>`
`GET /data/magnet/<URI-encoded magnet link>/<display path of file declared in torrent info>`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -56,13 +56,17 @@ Routes
...
--	`/events?ih=<infohash in hex>`. This is a websocket that emits frames with [confluence.Event] encoded as JSON for the torrent. The PieceChanged field for instance is set if the given piece changed [state](https://godoc.org/github.com/anacrolix/torrent#PieceState) within the torrent.
--	`GET /fileState?ih=<infohash in hex>&path=<display path of file declared in torrent info>`. Returns [file state](https://godoc.org/github.com/anacrolix/torrent#File.State) encoded as JSON.
--	`POST /metainfo?ih=<infohash in hex>`. The request body is a bencoded metainfo, as typically appears in a `.torrent` file. The trackers and info bytes are applied to the torrent matching the info hash provided in the query. No fields in the metainfo are mandatory.
--	`GET /metainfo?ih=<infohash in hex>`. returns a .torrent file containing the hash info.
+- `GET /data/infohash/<infohash in hex>/<display path of file declared in torrent info>`
+
+  `GET /data/magnet/<URI-encoded magnet link>/<display path of file declared in torrent info>`
```
Comment (anacrolix) [inline README.md:61]: This brings a whole new consistency problem to the library. Existing libraries unpack magnet links and update the torrent state with the magnet parameters by posting to /metainfo. If you accept magnet links here, you probably want to accept them for any handler that takes an infohash.

File: confluence/mux.go [hunk 2]
```diff
@@ -9,17 +9,17 @@ import (
 func (h *Handler) initMux() {
 	h.initMuxOnce.Do(func() {
 		mux := &h.mux
-		mux.Handle("/data", h.withTorrentContext(dataHandler))
+		mux.Handle("/data/", h.withTorrentContext(dataHandler))
```
Comment (anacrolix) [inline confluence/mux.go:12]: Doens't this prevent accepting `/data`, or does it redirect it?

Comment (ftk) [reply confluence/mux.go:12]: It seems to 301 redirect /data?... to /data/?... for some reason.

File: README.md (hunk 1 above)
Comment (ftk) [reply README.md:61]: Don't they already support `?magnet=<magnet URI>`? 
`.../magnet/...` and `.../infohash/...` is also supported for any torrent handler, didn't want to clutter the readme.

Comment (anacrolix) [reply README.md:61]: You're right, I have too many projects 😿 .

File: confluence/mux.go (hunk 2 above)
Comment (anacrolix) [reply confluence/mux.go:12]: That's behaviour of the standard lib http.ServeMux I think. There are a lot of projects already depending on the existing behaviour, I don't think I want to change this in the short-term.

**Comment by anacrolix** [comment]
@boramalper what do you think of this PR?

**Comment by boramalper** [comment]
I haven't personally tested it, but looks good to me. 👍

File: confluence/mux.go (hunk 2 above)
Comment (anacrolix) [reply confluence/mux.go:12]: This will need to be addressed to consider merging.

**Review by anacrolix — CHANGES_REQUESTED** [review]

**Comment by anacrolix** [comment]
I might make a few changes on top of the PR, let me know your thoughts on the changes. Sorry for the pedanticism but this is used heavily and very specifically by a lot of production services.

**Comment by anacrolix** [comment]
@ftk what do you think of https://github.com/anacrolix/confluence/pull/34? It refactors some stuff to make the changes in this PR fit in more in the style of this module, and makes it easier to adapt to other routes down the track (including potentially magnet and host name handling).

**Comment by ftk** [comment]
Thanks! I'll close this in favor of #34.

**Pull request closed without merging**